k8s_service: add new kubernetes module for handling Services

mmazur opened the pull request:
##### SUMMARY
Currently kubernetes is supported by the `k8s` module, which largely just takes an appropriately structured yaml and passes it to the kubernetes cluster for processing. This approach means that the one single module can handle anything kubernetes handles. The downside is that it's a rather un–ansibl…y? (–blonic?) approach with multi–level yaml required to do even the most basic stuff.

This PR adds a `k8s_service` module dedicated to handling the popular Service resource type of kubernetes. Its purpose is to answer the following questions:
1. Is ansible interested in having dedicated modules for the most popular kubernetes resource types.
2. If yes, how to go about implementing them.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
modules/clustering/k8s/k8s_service

##### ADDITIONAL INFORMATION
One implementation detail I'm not clear on is – the base KubernetesRawModule class provides/requires the existence of `kind`, `resource_definition` and `src` parameters. The first one should obviously not be an option in this case. The latter two aren't my preferred approaches to handling the issue, since instead of `resource_definition` I've added `spec` and `metadata` as options of the module in order to skip a level of configuration. It's currently done in a hacky way (hence the WIP tag), since I'd first like to get some confirmation on whether `spec:` and `metadata:` seem like the correct approach before spending more time on forcing `resource_definition` out.

willthames commented:
Multi-yaml documents are an implementation detail - you could just run a k8s module for each resource in turn, each resource having its own YAML file.

I can see that the interface for this module is definitely more Ansiblish, but I worry that providing this will also take us down a slippery slope of then suggesting that we need 100+ modules for each different core Kubernetes resource type (and then there's the custom resource definitions)

My vote on this would be to not do this - our time would be better spent improving the existing module to fill in missing gaps.

fabianvf commented:
First of all, thanks for the contribution, glad we're starting this conversation!

> I can see that the interface for this module is definitely more Ansiblish, but I worry that providing this will also take us down a slippery slope of then suggesting that we need 100+ modules for each different core Kubernetes resource type (and then there's the custom resource definitions)

I actually think this is a great pattern for the most popular core resources - it allows us to expose a more structured, focused interface with better documentation and validation. I definitely think we should do this in a limited fashion, but I don't think having a small set of modules to cover the most common workloads necessarily implies that we need to expand that to the whole API, especially since we always have the k8s module to fall back to.

> My vote on this would be to not do this - our time would be better spent improving the existing module to fill in missing gaps.

I noticed the implementation of this modules relied heavily on the existing code from our module_utils (a big +1 for that), and I think as long as we keep structuring it to build off the generic module we shouldn't have to worry too much about a split effort.

On to some commentary about the PR itself:

> I'd first like to get some confirmation on whether spec: and metadata: seem like the correct approach

If we're going to make specialized modules for specific resources, I think we should try to make them as native feeling as possible. I don't know if we gain a ton by tossing out `definition` just to pop one layer in. My feeling is that we should take what we want from spec/metadata and expose/document them in an ansible-y way directly in the module - if the user has something different that they're trying to do they can always just toss their definition into the `k8s` module.

mmazur commented:
The rationale for the existence of this module is basically UX – to have more benefit of using ansible for k8s control by getting simplified syntax and some config verification. This is also the reasoning for why modules for CRDs [already exist](https://github.com/kubevirt/ansible-kubevirt-modules/tree/master/lib/ansible/modules/clustering/kubevirt), even if they do not end up being mergeable.

And since this is about UX, it can't be an all or nothing decision, but a balance of what makes life easier in the most common situations. I do not have enough k8s experience to draw up a list of all the resource types that should have dedicated modules, but I'm 100% sure Service is on that list. And as Fabian noticed, this thing relies heavily on KubevirtRawModule, which should make any additional maintenance burden almost non–existent. Plus once we'd get to a third module like it, the amount of common code extracted would mean the modules themselves would be mostly parameter definitions and barely any code.

As for the module itself – my goal was to have the most common cases handled by the module directly while still making it easy to pass any additional parameters, as I do not think it a good idea to try to support everything, since that's like 30 parameters total: [meta](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#objectmeta-v1-meta), [spec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#servicespec-v1-core). Hence `spec:` and `metadata:`. Granted the same thing is achievable with `definition:`, but I just found that one level less looking nicer.

Anyway, for now I'll drop `spec:` and `metadata:`, add some minor fixups I have queued up and look into all the issues the bot flagged.

fabianvf commented:
> As for the module itself – my goal was to have the most common cases handled by the module directly while still making it easy to pass any additional parameters, as I do not think it a good idea to try to support everything, since that's like 30 parameters total: meta, spec. Hence spec: and metadata:. Granted the same thing is achievable with definition:, but I just found that one level less looking nicer.

Yeah, I understand this feeling (and agree that `spec` + `metadata` are better than `definition` for a more specialized module). I don't think having a bunch of parameters is necessarily a deal breaker, the configuration is there anyway so they're either there implicitly or explicitly. I think we could take all the most common arguments and bring them to the top so that we can document and validate them in a normal Ansible way. We can leave spec/metadata in there as well to cover our bases, but I think the ideal is that an average user going to create a service will not need to fall back to using those fields.

mmazur committed:
k8s: add k8s_kind arg to KubernetesRawModule

Single–kind k8s modules (e.g. k8s_service) do not have a module
parameter called 'kind' and need to pass a static 'kind' on
KubernetesRawModule class creation. Hence this change.

mmazur committed:
k8s: make 'validate' and 'wait' mod params optional

Not all k8s modules utilizing KubernetesRawModule will use these.

mmazur commented:
Ok, I'm done with all my fixes, all the tests have passed. Can this be merged?

fabianvf commented:
Since this module will set the pattern for all future specialized Kubernetes modules, I think we should discuss a little bit more what we want the interface to look like, and how much utility we want to provide for the user. For example, there are a bunch of specific annotations for interacting with a variety of cloud providers when creating a `LoadBalancer` type (https://kubernetes.io/docs/concepts/services-networking/service/#internal-load-balancer) - we could add a parameter that lets the user specify what cloud provider they're using and we could figure out the proper annotation for them. I also think we could pull some more fields from `spec` and put them in at the top level, there are only ~13 total for services I think.

This doesn't necessarily need to block a merge (we could add this functionality iteratively), but I would like the discussion to begin at least, since with a very specific module like this we can offer a lot of assistance to the users and really provide a dramatic improvement to UX for certain use-cases.

Anyone have thoughts on these higher-level sort of parameters in specialized modules like this? One issue I see is that parameters that deal with cloud providers may be difficult to impossible to test well.

mmazur commented:
My goals for this module were:
1. Get the minimal set of functionality that (a) I'm actually able to test and (b) sets the structure for how this module is to be used and further expanded.
2. Make sure `resource_definition` works, so that it's possible to define an arbitrarily complex Service right from the start.
3. Over time see more arguments being added, so that eventually nobody really needs to use `resource_definition` for anything (though keep it just to be safe).
4. UX is king.

Regarding that last one – the `k8s` module allows one to define `name` and `namespace`, but treats them as optional, since one defines stuff in `resource_definition` anyway, so why force the user to define those two as mod args. In `k8s_service` I intentionally did the opposite and marked those as `required` to get that early config checking and not have to wait for the cluster to complain of lacking arguments. Ditto with `type` argument, which has the allowed choices hardcoded in (not 100% on this one though; don't know enough about k8s and whether there aren't distros of it out there that have more Service types).

My point being that your idea with adding some kind of `cloud_provider` arg and using that to abstract away/generalize other stuff is perfectly in line with what I have intended for this to look like going forward[1]. And yes, it'd be a problem to test, but I'm not 100% sure that's such an obstacle. It's easy to verify whether all the supported args actually produce k8s config according to spec. And according to spec == working. :)

[1] This is also how the kubevirt ansible team approaches creating modules to handle our CRDs. Abstracting away the underlying complexity into something easier to use without cutting away functionality.

(In other news, our team produced a number of unit tests for our own CRD modules, that in practice cover most of the `k8s` module. Once this PR gets merged and we've settled on this approach of doing per–type modules, we'll likely clean up our tests and submit them for merging, so that it becomes easy to verify whether all those generalized arguments produce k8s configs conformant to the spec.)

mmazur commented:
Ping. I'm new to this, so not sure how long PRs are left to marinate before further steps are taken… :)

fabianvf commented on lib/ansible/modules/clustering/k8s/k8s_service.py:
```diff
@@ -0,0 +1,285 @@
+}
+
+
+def virtdict():
```
What does this do?

mmazur replied:
It allows me to do this, which makes the whole 'preparing object config code' easy to read:

```
definition = virtdict()
def_meta = definition['metadata']
def_meta['name'] = 'asdf'
def_labels = def_meta['labels']
def_labels['key'] = 'special'
(…)
```

And it works to an infinite nesting level. Granted, I don't need infinite, so if you don't like the dangling global function, I can do two things:
1. Embed `def virtdict()` inside `execute_module()`.
2. Drop `virtdict` and switch to using `definition = defaultdict(defaultdict)`, which gives me one level, but for the current code one level is enough.

Any preference?

mmazur replied:
Picked option 2.

fabianvf commented on lib/ansible/modules/clustering/k8s/k8s_service.py:
```diff
@@ -0,0 +1,285 @@
+    - strategic-merge
+    type: list
+  api_version:
+    description:
```
Do we still need kind? For service I think it should always be v1

mmazur replied:
Agreed, at present this serves no purpose. If a v2 Service ever comes into being, we'll likely need to bring the arg back, but we can deal with the issue if and when that happens. I'll drop it.

mmazur replied:
Removed.

fabianvf commented on lib/ansible/modules/clustering/k8s/k8s_service.py:
```diff
@@ -0,0 +1,285 @@
+    },
+    'name': {'required': True},
+    'namespace': {'required': True},
+    'api_version': {
```
Same comment as above

mmazur replied:
Removed.

fabianvf commented on lib/ansible/modules/clustering/k8s/k8s_service.py:
```diff
@@ -0,0 +1,285 @@
+            'NodePort', 'ClusterIP', 'LoadBalancer', 'ExternalName'
+        ],
+    },
+    'ports': {'type': 'list'},
```
@willthames can we specify the structure of the why elements of this list in this spec?

fabianvf reviewed (commented):
Thanks for the ping, sorry for letting that sit. Had a couple more questions but other than those looks good to me.

mmazur commented:
Ping.

fabianvf commented:
shipit

mmazur committed:
k8s_service: new k8s module for handling Services

maxamillion commented:
shipit

mmazur commented:
There was an error in `name` and `namespace` parameter description. I've pushed a corrected version. Dunno how 'shipit' works, so I'm pinging in case someone needs to invoke the magic spell once more.

maxamillion commented:
@fabianvf nvm :)

fabianvf commented:
Heh, well it looked good to me anyway
